### PR TITLE
Params of type boolean should be coersed to null if string value is n…

### PR DESCRIPTION
**Pull request by rthunbo**

### CODE REVIEW

**Comment by jsdevel** [comment]
can you add tests?  I'd like to see the use cases.

**Comment by jsdevel** [comment]
Thanks for adding a test!  I see what you mean now.

So I think the value should be coerced into a `boolean`.  Coercing to `null` could lead to `TypeError` being thrown.  What is the current behavior?

**Comment by rthunbo** [comment]
Well if you always coerce boolean query params to a boolean then express-openapi-validation will always succeed on boolean query params no matter if the query param value is "true", "false", "bogus". 
Therefore I think the coercion should return null for the "bogus" value, so that the validation fails in this case. 
Otherwise there is no reason to have any kind of validation on boolean query params, as it would always be coerced to either true or false before any validation takes place.
The change I made I have tested and due to the validation now firing you will not reach the controller, of cause if you don't use validation there can be an issue, but that would be due to erroneous input anyway. And for numbers you already coerce them to NaN if you input "bogus" in that case.
I think you should also need to consider the usecase with using express-openapi-validation.

**Comment by jsdevel** [comment]
One concern I have is that this is a breaking change.

I also don't believe coercion middleware should be concerned about validation.  In theory this could be used without validation middleware in the middleware chain.

**Comment by jsdevel** [comment]
Is there a jsonschema directive for allowing null?  Perhaps we could coerce to null if something like that were present and otherwise follow the original behavior to avoid a breaking change?

**Comment by rthunbo** [comment]
Wouldn't help though. There are an allowEmptyValue but by setting this then the validation wouldn't fail. So we're back to there not being any possibility of having validation fail on any boolean query parameter, it will let anything through, because there is no way for the validation library to validate prior to that because it has already been coersed.

**Comment by jsdevel** [comment]
@rthunbo if you're ultimately concerned about validation, then it's the ordering of middleware that likely needs to change.  Can you see if there's a way to change the ordering of the middleware in `express-openapi`?  If validation middleware occurred prior to coercion middleware, would that solve your issue?

**Comment by rthunbo** [comment]
No unfortunate not. The validation middleware expect the correct types, so if the coersion middleware is disabled, the validation will fail on any input. I have made the workaround that in the route the query param is set to "string" accepting only "true" and "false", and then coercing that value to boolean inside the controller.

**Comment by rthunbo** [comment]
I have changed the implementation to use a vendor extension. Please have a look

**Comment by rthunbo** [comment]
@jsdevel I have changed the implementation to use a vendor extension. Please have a look.

File: index.js [hunk 1]
```diff
@@ -97,6 +113,8 @@ function buildCoercer(params, isHeaders) {
       var coercer;
       var itemCoercer;
       var type = param.type;
+	  
+      var strict = param["x-express-openapi-strict-coercion"];
```
Comment (jsdevel) [inline index.js]: Can you change this to `x-express-openapi-coercion-sctrict`?

**Comment by rthunbo** [comment]
@jsdevel done.

**Comment by rthunbo** [comment]
@jsdevel I have implemented the desired change. When do you think this can be merged in?

File: index.js [hunk 2]
```diff
@@ -137,8 +155,15 @@ function byLocation(location) {
   };
 }
 
-function getCoercer(type) {
-  return COERCION_STRATEGIES[type];
+function getCoercer(type, strict) {
+  var strategy = undefined;
```
Comment (jsdevel) [inline index.js]: can you remove ` = undefined`?

**Comment by rthunbo** [comment]
@jsdevel Done.

**Comment by jsdevel** [comment]
Thanks for being patient @rthunbo !  I like this approach a lot and feel it will be a great addition.

**Comment by jsdevel** [comment]
Can you add this to the README documentation?

**Pull request merged**